### DISCUSSION TITLE
Page and assets matching for service worker precache (PWA)

### DISCUSSION THREAD

**Discussion by zacBkh** [vercel/next.js#64336] [Help]
### Summary

I am working on a PWA powered by NextJS but having hard time to tell the cache API which resource to cache because of how files are served (bundling/webpack/chunks...)

### Additional information

Hi, everyone
I am Zacharie, a french dev.

Recently I had to learn about service worker for a project I have, which requires a specific page of my app to be always accessible offline.
I managed to build a small playground website where a specific page is always available offline.

However, my real world project is more complicated and has business logic, so I am building it with NextJS App Router (14.1.4)
The big difference between a front end framework like Next and a traditional website built with .html files is that Next serve the assets through build-generated files (usually .js files) and those are very cryptic and their name is randomly generated at each build I belive.
NextJS also uses Webpack/Turbopack.

My big issue is that I don't know how to select the resources to cache, because next js is using an abstraction that I don't fully understand when delivering the assets.

To cache, I would typically do a (inside an `installEventListener`) :
```js
 evt.waitUntil(
    caches.open(CACHE_NAME).then((cache) => {
      cache.addAll(CACHE_ASSETS)
    })
  )
```
In a basic website, CACHE_ASSETS could be : "offline.html"
But in my next js app I don't know at all what to put, since Next provides an abstraction when delivering the files.

Naively, I tried to look for the resource requested for each page to load in `.next\app-build-manifest.json` (which is kind of a lookup table page<>resources) but it's not working and I don't think it is the ideal solution to rely on those build-generated files.

Still, in my nextJS app, I manage to have offline support when I add a `offline.html` page in my `/public folder`, but that's not the way I want to build the app...
```

**Comment by Link2Twenty** [comment]
This is also an issue I faced recently. I solved it by manually caching the required files on first load from JS. 

1. Cache the required page
2. Parse all newly cached HTML and CSS files to find required files and assets, if there are none stop
3. Cache files found in parsing
4. GOTO step 2

This feels a little hacky but there isn't anything I could see that allows me to see what files and assets are required for any give route to load.
